### PR TITLE
[RISCV] Enable load clustering by default

### PR DESCRIPTION
**Pull request by asb**
Stacks on top of #75338 and #75341

Posting for comment and for people to test on their workloads, feedback
on ideas for a tweaked heuristic etc.

### CODE REVIEW

**Review by preames — COMMENTED** [review]
Looking at the test diffs, two observations:
1) This appears to be increasing register pressure in some cases.  I can't find any case where it causes a spill it didn't before, but we do have cases where more registers are live at the same time.
2) Clustering doesn't appear to be sorting by offset, and instead appears to be a somewhat arbitrary reordering of already adjacent loads.  (When they happen to already be adjacent.)

Does this work if we try doing this post RA?  That would cut down on the potential register pressure concern.  To be clear, this is just a question, I could easily be convinced that doing this pre-RA is the right answer.  

Your heuristic looks reasonable to me.  I don't have data one way or another to indicate a preference for it, or anything else.

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp [hunk 1]
```diff
@@ -2281,9 +2282,9 @@ bool RISCVInstrInfo::shouldClusterMemOps(
     return false;
   }
 
-  // TODO: Use a more carefully chosen heuristic, e.g. only cluster if offsets
-  // indicate they likely share a cache line.
-  return ClusterSize <= 4;
+  // A cache line is typically 64 bytes, so cluster if the memory ops are on
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:2285]: We can get cache line size via `TargetTransformInfo::getCacheLineSize` or `MCSubtargetInfo::getCacheLineSize` (for the default implementation, `TargetTransformInfo::getCacheLineSize` just calls `MCSubtargetInfo::getCacheLineSize`). If the return value is not 0, we can use it as a more exact value.
For RISCV, the cache line size info can be specified when defining processor.

**Comment by asb** [comment]
> Looking at the test diffs, two observations:
> 
>     1. This appears to be increasing register pressure in some cases.  I can't find any case where it causes a spill it didn't before, but we do have cases where more registers are live at the same time.
> 
>     2. Clustering doesn't appear to be sorting by offset, and instead appears to be a somewhat arbitrary reordering of already adjacent loads.  (When they happen to already be adjacent.)
> 
> 
> Does this work if we try doing this post RA? That would cut down on the potential register pressure concern. To be clear, this is just a question, I could easily be convinced that doing this pre-RA is the right answer.
> 
> Your heuristic looks reasonable to me. I don't have data one way or another to indicate a preference for it, or anything else.

Digging into the ordering behaviour, it is indeed a bit confusing. BaseMemOpClusterMutation::apply will collect the MemOpInfo (calling getMemOperandsWithOffsetWidth for this), then they're put into group (commonly one big group, unless over a certain size threshold), and then the array of MemOpInfo is sorted into ascending order based on base+offset. clusterNeighbouringMemOps then performs clustering using that order __except__ there's logic that avoids reordering that was added in <https://reviews.llvm.org/D72706> with the argument that the reordering "leads to test case churn and 'goes against the machine schedulers "do no harm" philosophy'. I'm not sure I follow how the reordering could be harmful, and it does seem unfortunate because it seems to me the sdag scheduler does do this reordering so there's even more churn than you'd think when trying to check the incremental benefit of adding sdag load clustering in addition.

I've added a new parameter to the pass that controls this behaviour, and enabled the reordering for RISC-V. If people are happy with this direction, I'd intend to split this out as a patch that this one depends on.

Trying to enable this for the post-ra scheduler it seems to fail to make any changes (i.e. doesn't work), though it's possible I made an error.

Thanks @wangpc-pp for the suggestion re querying cache line size. I've not added that in this iteration, but will do so in the next rev.

File: llvm/include/llvm/CodeGen/TargetInstrInfo.h [hunk 2]
```diff
@@ -1496,12 +1496,18 @@ class TargetInstrInfo : public MCInstrInfo {
...
+  /// \p Offset1 and \p Offset2 are the byte offsets for the memory
+  /// operations.
+  /// \p OffsetIsScalable1 and \p OffsetIsScalable2 indicate if the offset is
+  /// scaled by a runtime quantity.
   /// \p ClusterSize is the number of operations in the resulting load/store
   /// cluster if this hook returns true.
   /// \p NumBytes is the number of bytes that will be loaded from all the
   /// clustered loads if this hook returns true.
   virtual bool shouldClusterMemOps(ArrayRef<const MachineOperand *> BaseOps1,
+                                   int64_t Offset1, bool OffsetIsScalable1,
```
Comment (zixuan-wu) [inline llvm/include/llvm/CodeGen/TargetInstrInfo.h:1508]: Do we need a width operand too?

Comment (asb) [reply llvm/include/llvm/CodeGen/TargetInstrInfo.h]: You could imagine trying to use that, but our current heuristic doesn't make use of that information so I don't think it's necessary to expand the parameter list further.

**Comment by asb** [comment]
I've rebased and split out two patches for separate review (#75338 and #75341) in order to hopefully inch this forward. The rebase will also pick up some new test changes due to #73802 landing.

File: llvm/lib/Target/RISCV/RISCVTargetMachine.cpp [hunk 3]
```diff
@@ -95,11 +95,6 @@ static cl::opt<bool>
                         cl::desc("Enable Split RegisterAlloc for RVV"),
                         cl::init(true));
 
-static cl::opt<bool> EnableMISchedLoadClustering(
-    "riscv-misched-load-clustering", cl::Hidden,
-    cl::desc("Enable load clustering in the machine scheduler"),
-    cl::init(false));
```
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVTargetMachine.cpp:101]: In this patch, please just flip the bool default here.  If you want to remove the option entirely, do that in a separate change in a couple weeks.

**Review by preames — APPROVED** [review]
LGTM w/comment addressed.  The comment is basically a rework of the patch, but the resulting patch is near trivial so it felt reasonable to do a conditional LGTM here.

**Comment by wangpc-pp** [comment]
Reverse ping!
@asb We want to enable this along with store clustering. We can make them subtarget features if there are objections.

**Comment by asb** [comment]
Thanks for the ping, I've been intending to return to this so it's good to hear there's interest. I'll rebase in the coming days and try to move this forwards.

**Comment by asb** [comment]
I've rebased and repushed (again actually, I had done it before but forgotten to push). I've adopted Philip's suggestion of just flipping the default for the option for this patch.

File: llvm/test/CodeGen/RISCV/srem-vector-lkk.ll [hunk 4]
```diff
@@ -1085,17 +1085,18 @@ define <4 x i64> @dont_fold_srem_i64(<4 x i64> %x) nounwind {
...
+; RV32I-NEXT:    lw a3, 0(a1)
+; RV32I-NEXT:    lw a4, 4(a1)
 ; RV32I-NEXT:    lw s4, 8(a1)
 ; RV32I-NEXT:    lw s5, 12(a1)
-; RV32I-NEXT:    lw a3, 0(a1)
-; RV32I-NEXT:    lw a1, 4(a1)
 ; RV32I-NEXT:    mv s6, a0
 ; RV32I-NEXT:    li a2, 1
 ; RV32I-NEXT:    mv a0, a3
+; RV32I-NEXT:    mv a1, a4
```
Comment (wangpc-pp) [inline llvm/test/CodeGen/RISCV/srem-vector-lkk.ll:1099]: Extra move?

File: llvm/test/CodeGen/RISCV/srem-vector-lkk.ll [hunk 5]
```diff
@@ -1154,17 +1155,18 @@ define <4 x i64> @dont_fold_srem_i64(<4 x i64> %x) nounwind {
...
+; RV32IM-NEXT:    lw a3, 0(a1)
+; RV32IM-NEXT:    lw a4, 4(a1)
 ; RV32IM-NEXT:    lw s4, 8(a1)
 ; RV32IM-NEXT:    lw s5, 12(a1)
-; RV32IM-NEXT:    lw a3, 0(a1)
-; RV32IM-NEXT:    lw a1, 4(a1)
 ; RV32IM-NEXT:    mv s6, a0
 ; RV32IM-NEXT:    li a2, 1
 ; RV32IM-NEXT:    mv a0, a3
+; RV32IM-NEXT:    mv a1, a4
```
Comment (wangpc-pp) [inline llvm/test/CodeGen/RISCV/srem-vector-lkk.ll:1169]: And here.

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-fp.ll [hunk 6]
```diff
@@ -889,170 +889,194 @@ define void @copysign_v8f16(ptr %x, ptr %y) {
...
 ; ZVFHMIN-ZFHIN-RV32-LABEL: copysign_v8f16:
 ; ZVFHMIN-ZFHIN-RV32:       # %bb.0:
-; ZVFHMIN-ZFHIN-RV32-NEXT:    addi sp, sp, -32
-; ZVFHMIN-ZFHIN-RV32-NEXT:    .cfi_def_cfa_offset 32
+; ZVFHMIN-ZFHIN-RV32-NEXT:    addi sp, sp, -48
+; ZVFHMIN-ZFHIN-RV32-NEXT:    .cfi_def_cfa_offset 48
+; ZVFHMIN-ZFHIN-RV32-NEXT:    sw s0, 44(sp) # 4-byte Folded Spill
+; ZVFHMIN-ZFHIN-RV32-NEXT:    sw s1, 40(sp) # 4-byte Folded Spill
+; ZVFHMIN-ZFHIN-RV32-NEXT:    sw s2, 36(sp) # 4-byte Folded Spill
+; ZVFHMIN-ZFHIN-RV32-NEXT:    sw s3, 32(sp) # 4-byte Folded Spill
```
Comment (wangpc-pp) [inline llvm/test/CodeGen/RISCV/rvv/fixed-vectors-fp.ll:897]: Is this because the register pressure increases?

**Comment by wangpc-pp** [comment]
> I've rebased and repushed (again actually, I had done it before but forgotten to push). I've adopted Philip's suggestion of just flipping the default for the option for this patch.

Thanks! It seems there are new conflicts (we are actively changing).
I found some regressions, but I don't mean to block this patch.

**Comment by asb** [comment]
> > I've rebased and repushed (again actually, I had done it before but forgotten to push). I've adopted Philip's suggestion of just flipping the default for the option for this patch.
> 
> Thanks! It seems there are new conflicts (we are actively changing). I found some regressions, but I don't mean to block this patch.

Easy enough for me to regenerate and fix the conflicts if we're happy to merge. As you point out, there are some minor regressions in a few tests but I think what I need from reviewers is an indication that they agree this is the right direction, and if we find any regressions that are meaningful in practice we can follow up with fixes in tree. I can do my own testing, but given your ping that you would like to see this enabled, I'm guessing you've observed it's at least no worse than the status quo on benchmarks, which would be supportive of moving ahead?

**Review by wangpc-pp — APPROVED** [review]
LGTM! Thanks a lot!
We have already enabled it internally, and we do see some improvements on LMBench.

**Comment by preames** [comment]
Please land this.

**Pull request merged**